packages/gatsby/src/utils/webpack-utils.ts

```diff
@@ -769,9 +769,25 @@ export const createWebpackUtils = (
     }
   ): CssMinimizerPlugin => new CssMinimizerPlugin(options)
 
-  plugins.fastRefresh = (): Plugin =>
+  plugins.fastRefresh = ({ modulesThatUseGatsby }): Plugin =>
     new ReactRefreshWebpackPlugin({
       overlay: false,
+      // this is a bit hacky - exclude expect string or regexp or array of those
+      // so this is tricking ReactRefreshWebpackPlugin that we provide RegExp
+      exclude: {
+        test: modulePath => {
+          // when it's not coming from node_modules we treat it as a source file.
+          if (!vendorRegex.test(modulePath)) {
+            return false
+          }
+
+          // If the module uses Gatsby as a dependency
+          // we want to treat it as src because of shadowing
+          return !modulesThatUseGatsby.some(module =>
+            modulePath.includes(module.path)
+          )
+        },
+      } as RegExp,
```

Comment:
This feels very hacky and if we would go with this - I'd probably suggest pinning `@pmmmwh/react-refresh-webpack-plugin` to exact version as this only implement `test` method of RegExp and this might break spectacularly if other methods would be used in future versions.

Ideally we could ask for support of function for `exclude`, but this is also weird because that webpack plugin use helpers provided by webpack ( https://github.com/webpack/webpack/blob/04a7d052b5c04628257ec23244f234622c7f394c/lib/ModuleFilenameHelpers.js#L230-L257 ) which just don't handle functions (weird part to me is that rules do, so why would that helper not support is a bit baffling)